File: src/infra/llm/claude-cli.ts
import { spawn as defaultSpawn, type ChildProcess } from 'node:child_process'
import { createInterface } from 'node:readline'
import type { LLMProvider, InvokeOptions, StreamEvent, LLMMode } from './types.js'
import { parseClaudeStreamLine } from '../../ui/stream-parser.js'

export type SpawnFn = typeof defaultSpawn

export interface ClaudeCliConfig {
  binary?: string
  timeoutMs?: number
  spawn?: SpawnFn  // injectable for testing
}

export type ExitVerdict =
  | { kind: 'success' }
  | { kind: 'oauth-needed' }
  | { kind: 'generic-error' }

export class ClaudeCliProvider implements LLMProvider {
  private child: ChildProcess | null = null
  private cfg: Required<ClaudeCliConfig>

  constructor(cfg: ClaudeCliConfig = {}) {
    this.cfg = {
      binary: cfg.binary ?? 'claude',
      timeoutMs: cfg.timeoutMs ?? 30 * 60 * 1000,
      spawn: cfg.spawn ?? defaultSpawn
    }
  }

  // Three must-set sandbox flags (spike-verified):
  //  - acceptEdits: default mode + -p blocks all Write tool calls (spike B);
  //    acceptEdits auto-accepts Write/Edit while Bash etc still ask.
  //  - allowedTools: WHITELIST instead of blacklist. Anything not in the
  //    list is auto-denied. Forward-compat: new Claude Code tools
  //    (NotebookEdit / Task / AskUserQuestion / MCP servers / future
  //    additions) cannot silently enter the agent's toolbox without an
  //    explicit codebus update. Specifically protects against
  //    AskUserQuestion which would hang the -p flow waiting for input
  //    that has no terminal to come from. Spec §16 listed this as a
  //    phase 2 evaluation; landed in phase 1 after manual test surfaced
  //    the leak.
  //  - cwd is supplied via opts.cwd in invoke(): spike E confirmed cwd =
  //    .codebus/ gives system-level isolation from user source repo.
  // No --add-dir: spike confirmed it widens, not narrows; cannot scope cwd.
  buildArgv(opts: { mode: LLMMode; vaultRoot: string }): string[] {
    // Read/Glob/Grep are universal (both modes need to discover + read).
    // Write/Edit only in ingest — query is hard read-only.
    const allowed = ['Read', 'Glob', 'Grep']
    if (opts.mode === 'ingest') allowed.push('Write', 'Edit')
    void opts.vaultRoot  // kept in signature for phase 2 settings whitelist
    return [
      '-p',
      '--output-format', 'stream-json',
      '--input-format', 'stream-json',
      '--verbose',
      '--permission-mode', 'acceptEdits',
      '--allowedTools', allowed.join(',')
    ]
  }

  async *invoke(opts: InvokeOptions): AsyncIterable<StreamEvent> {
    const argv = this.buildArgv({ mode: opts.mode, vaultRoot: opts.vaultRoot })
    this.child = this.cfg.spawn(this.cfg.binary, argv, { cwd: opts.cwd })

    const timer = setTimeout(() => this.cancel(), this.cfg.timeoutMs)

    let stderrBuf = ''
    this.child.stderr?.on('data', (chunk) => { stderrBuf += chunk.toString() })

    // Single user-turn message via stream-json input. Real schema: top-level
    // {type:"user", message:{role:"user", content:"..."}}.
    const inputMsg = {
      type: 'user',
      message: { role: 'user', content: `${opts.systemPrompt}\n\n${opts.userMessage}` }
    }
    this.child.stdin?.write(JSON.stringify(inputMsg) + '\n')
    this.child.stdin?.end()

    const stdout = this.child.stdout
    if (!stdout) throw new Error('claude -p produced no stdout')
    const rl = createInterface({ input: stdout })

    try {
      for await (const line of rl) {
        if (!line.trim()) continue
        for (const event of parseClaudeStreamLine(line)) yield event
      }
      const code: number = await new Promise((resolve) =>
        this.child!.once('exit', (c) => resolve(c ?? 0))
      )
      const verdict = this.classifyExit(code, stderrBuf)
      if (verdict.kind === 'oauth-needed') {
        throw new Error(
          'Claude CLI 未認證 — 請在 terminal 跑 `claude` 完成 OAuth，再重新執行 codebus'
        )
      }
      if (verdict.kind === 'generic-error') {
        throw new Error(`claude -p exited ${code}: ${stderrBuf.slice(0, 500)}`)
      }
      yield { kind: 'done' }
    } finally {
      clearTimeout(timer)
    }
  }

  classifyExit(code: number, stderr: string): ExitVerdict {
    if (code === 0) return { kind: 'success' }
    if (/unauthen|auth(?:enticat)?(?:ed|ion)?|token|login/i.test(stderr)) {
      return { kind: 'oauth-needed' }
    }
    return { kind: 'generic-error' }
  }

  cancel(): void {
    if (this.child && !this.child.killed) {
      this.child.kill('SIGTERM')
    }
  }
}

File: codebus-core/src/llm/claude_cli.rs
//! `claude -p` subprocess provider — the Phase 1 [`LlmProvider`] impl.
//!
//! Sandbox semantics (iter-9 spike-verified, hard-won lesson):
//!
//! - `--permission-mode acceptEdits`: auto-accepts every tool in the toolset
//!   under `-p` mode (no terminal to deny prompts). Confirmed by manual test
//!   that Bash calls succeeded when the toolset still contained Bash, even
//!   under `--allowedTools='Read,Glob,Grep'`. So `acceptEdits` alone is NOT
//!   a sandbox.
//! - `--tools <list>`: WHITELIST. Tools not in this list are not visible to
//!   the agent at all. THIS is the lever that keeps Bash / WebFetch / future
//!   tools out. Iter-1 through iter-8 wrongly relied on `--allowedTools`.
//! - `--allowedTools <list>`: redundant safety net mirroring `--tools` so
//!   future Claude Code permission-mode changes don't accidentally hang on
//!   a prompt with no terminal.
//! - `cwd = .codebus/` (system-level isolation from user source repo,
//!   confirmed by spike E). No `--add-dir`: it widens, not narrows.

use crate::llm::provider::{EventStream, InvokeOptions, LlmMode, LlmProvider, ProviderError};
use crate::stream::{parse_claude_stream_line, StreamEvent};
use futures_util::StreamExt;
use std::path::Path;
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::Command;
use tokio_stream::wrappers::LinesStream;

/// Tools every mode allows: read-only discovery + read.
pub const ALWAYS_ALLOWED: &[&str] = &["Read", "Glob", "Grep"];

/// Additional tools allowed only in [`LlmMode::Ingest`] (writes wiki pages).
pub const INGEST_EXTRA: &[&str] = &["Write", "Edit"];

/// Tools that MUST never reach the agent under any mode. Sentinel set used
/// by the iter-9 negative assertion test — if any of these ever appears in
/// argv, the sandbox is broken.
pub const FORBIDDEN_TOOLS: &[&str] = &["Bash", "WebFetch", "WebSearch", "TodoWrite", "NotebookEdit", "BashOutput", "KillShell"];

/// Build the argv passed to `claude -p`. Pure function; tests pin both the
/// positive list (every allowed tool present) and the negative list (no
/// forbidden tool anywhere).
pub fn build_argv(mode: LlmMode, _vault_root: &Path) -> Vec<String> {
    let mut allowed: Vec<&str> = ALWAYS_ALLOWED.to_vec();
    if matches!(mode, LlmMode::Ingest) {
        allowed.extend_from_slice(INGEST_EXTRA);
    }
    let list = allowed.join(",");
    vec![
        "-p".into(),
        "--output-format".into(),
        "stream-json".into(),
        "--input-format".into(),
        "stream-json".into(),
        "--verbose".into(),
        "--permission-mode".into(),
        "acceptEdits".into(),
        "--tools".into(),
        list.clone(),
        "--allowedTools".into(),
        list,
    ]
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExitVerdict {
    Success,
    OauthNeeded,
    GenericError,
}

/// Classify the exit code + stderr buffer into one of three verdicts.
/// Mirrors TS `classifyExit`. The OAuth heuristic (regex over stderr) is
/// the same — if it ever drifts, both impls have to update together.
pub fn classify_exit(code: i32, stderr: &str) -> ExitVerdict {
    if code == 0 {
        return ExitVerdict::Success;
    }
    let lower = stderr.to_lowercase();
    if lower.contains("unauthen")
        || lower.contains("authentication")
        || lower.contains("authenticated")
        || lower.contains("authentic")
        || lower.contains("token")
        || lower.contains("login")
    {
        return ExitVerdict::OauthNeeded;
    }
    ExitVerdict::GenericError
}

/// Concrete [`LlmProvider`] backed by the `claude` CLI subprocess.
/// Sandbox argv built by [`build_argv`]; spawn cwd = `opts.cwd` provides
/// system-level isolation per spike E.
pub struct ClaudeCliProvider {
    binary: String,
}

impl ClaudeCliProvider {
    pub fn new() -> Self {
        Self { binary: "claude".into() }
    }

    pub fn with_binary(binary: impl Into<String>) -> Self {
        Self { binary: binary.into() }
    }
}

impl Default for ClaudeCliProvider {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait::async_trait]
impl LlmProvider for ClaudeCliProvider {
    async fn invoke(&self, opts: InvokeOptions) -> Result<EventStream, ProviderError> {
        let argv = build_argv(opts.mode, &opts.vault_root);

        let mut child = Command::new(&self.binary)
            .args(&argv)
            .current_dir(&opts.cwd)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true)
            .spawn()
            .map_err(|e| ProviderError::Setup {
                message: format!(
                    "Failed to spawn `{binary}`: {e}. Install Claude CLI from https://claude.ai/code and verify it is on PATH.",
                    binary = self.binary
                ),
            })?;

        let mut stdin = child.stdin.take().expect("stdin piped");
        let stdout = child.stdout.take().expect("stdout piped");

        // Build the single user-turn message. Real schema:
        // {type:"user", message:{role:"user", content:"..."}}.
        let payload = serde_json::json!({
            "type": "user",
            "message": {
                "role": "user",
                "content": format!("{}\n\n{}", opts.system_prompt, opts.user_message)
            }
        });
        let line = format!("{payload}\n");

        tokio::spawn(async move {
            let _ = stdin.write_all(line.as_bytes()).await;
            let _ = stdin.shutdown().await;
        });

        let line_stream = LinesStream::new(BufReader::new(stdout).lines());
        let event_stream = line_stream
            .flat_map(|line_result| {
                let events = line_result
                    .ok()
                    .map(|l| parse_claude_stream_line(&l))
                    .unwrap_or_default();
                futures_util::stream::iter(events)
            })
            .chain(futures_util::stream::once(async move {
                // Reap the child so it's not zombified. We ignore the exit
                // verdict here — Phase C goal/query commands inspect their
                // own provider stream and stderr separately to surface
                // OAuth / generic-error verdicts to the user. The Done
                // event terminates the stream cleanly regardless.
                let _ = child.wait().await;
                StreamEvent::Done
            }));

        Ok(Box::pin(event_stream))
    }

    fn cancel(&self) {
        // Cancellation is realized by dropping the consumed stream — the
        // chained `child.wait()` future drops with it, and the Child's
        // `kill_on_drop(true)` setting sends SIGTERM. If a future caller
        // needs an explicit out-of-band cancel handle, store the Child in
        // an Arc<Mutex<Option<Child>>> on self and SIGTERM here.
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    // === iter-9 sandbox argv tests — DOUBLE assertions per task 3.10 ===

    #[test]
    fn ingest_argv_contains_full_allowed_set() {
        let argv = build_argv(LlmMode::Ingest, &PathBuf::from("/v"));
        let joined = argv.join(" ");
        // Both --tools and --allowedTools mirror each other and contain
        // the same list of allowed names (iter-9 redundant safety net).
        assert!(joined.contains("--tools Read,Glob,Grep,Write,Edit"));
        assert!(joined.contains("--allowedTools Read,Glob,Grep,Write,Edit"));
    }

    fn flag_value<'a>(argv: &'a [String], flag: &str) -> &'a str {
        let pos = argv.iter().position(|a| a == flag).expect(flag);
        argv[pos + 1].as_str()
    }

    #[test]
    fn query_argv_excludes_write_edit() {
        let argv = build_argv(LlmMode::Query, &PathBuf::from("/v"));
        // Inspect the --tools / --allowedTools VALUES specifically — the
        // overall argv legitimately contains the substring "Edit" via
        // `acceptEdits` (the permission mode), which is unrelated to the
        // toolset whitelist.
        assert_eq!(flag_value(&argv, "--tools"), "Read,Glob,Grep");
        assert_eq!(flag_value(&argv, "--allowedTools"), "Read,Glob,Grep");
        let tools = flag_value(&argv, "--tools");
        assert!(!tools.contains("Write"));
        assert!(!tools.contains("Edit"));
    }

    #[test]
    fn forbidden_tools_never_appear_in_argv_under_any_mode() {
        // Iter-9 lesson hard-pinned: the sandbox MUST NOT leak any of these
        // tool names into any position in the argv (not as values to any
        // flag, not as positional). Anything in FORBIDDEN_TOOLS that ever
        // shows up here means the toolset whitelist regressed.
        for mode in [LlmMode::Ingest, LlmMode::Query] {
            let argv = build_argv(mode, &PathBuf::from("/v"));
            for forbidden in FORBIDDEN_TOOLS {
                for arg in &argv {
                    assert!(
                        !arg.contains(forbidden),
                        "forbidden tool {forbidden:?} found in argv for mode {mode:?}: {argv:?}"
                    );
                }
            }
        }
    }

    #[test]
    fn argv_uses_p_flag_with_stream_json_io() {
        let argv = build_argv(LlmMode::Query, &PathBuf::from("/v"));
        assert!(argv.contains(&"-p".to_string()));
        assert!(argv.contains(&"--output-format".to_string()));
        assert!(argv.contains(&"stream-json".to_string()));
        assert!(argv.contains(&"--input-format".to_string()));
        assert!(argv.contains(&"--verbose".to_string()));
    }

    #[test]
    fn argv_pins_permission_mode_to_accept_edits() {
        let argv = build_argv(LlmMode::Ingest, &PathBuf::from("/v"));
        let joined = argv.join(" ");
        assert!(joined.contains("--permission-mode acceptEdits"));
    }

    // === classify_exit tests ===

    #[test]
    fn exit_code_zero_is_success() {
        assert_eq!(classify_exit(0, ""), ExitVerdict::Success);
        assert_eq!(classify_exit(0, "irrelevant stderr"), ExitVerdict::Success);
    }

    #[test]
    fn auth_phrases_in_stderr_trigger_oauth_needed() {
        for stderr in [
            "Error: Unauthenticated request",
            "AUTHENTICATION required",
            "authentic mismatch",
            "no token found",
            "please login first",
        ] {
            assert_eq!(classify_exit(1, stderr), ExitVerdict::OauthNeeded, "stderr: {stderr}");
        }
    }

    #[test]
    fn unrelated_failure_is_generic_error() {
        assert_eq!(classify_exit(1, "Out of memory"), ExitVerdict::GenericError);
        assert_eq!(classify_exit(2, "syntax error"), ExitVerdict::GenericError);
        assert_eq!(classify_exit(127, "command not found"), ExitVerdict::GenericError);
    }
}
